### PR TITLE
Add cartridges overview in explorer

### PR DESCRIPTION
**Pull request by taurgis**
screenshot_cartridges

Started development on a cartridge overview. Still in development, so wouldn't recommend accepting the pull request just yet.

Just wanted to let you know what i'm up to.

### CODE REVIEW

**Comment by taurgis** [comment]
@SqrTT This is now mergeable (if it is ok for you). I have removed some unused code and refactored it. 

Also I have added some file icons and right clicking on a cartridge in the explorer will show a context menu to force a  clean/uploadall.

This is just a stepping stone to whats possible here off course: Later one it could be possible to add context menu's to add specific file types and put them in to the correct folder or even create a cartridge.

File: src/providers/CartridgesView.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+			return directories.map(toFolderElement).concat(files.map(toFileElement));
+		}
+
+		return [new CartridgeItem('No files', 'cartridge-file', '', vscode.TreeItemCollapsibleState.None)];
+	}
+
+
+	private getCartridgesInWorkspace(workspaceRoot: string): CartridgeItem[] {
+		if (this.pathExists(workspaceRoot)) {
+			var directories = this.getDirectories(workspaceRoot)
```
Comment (SqrTT) [inline src/providers/CartridgesView.ts]: Not always cartridges root and `workspaceRoot` is the same folder. For example in my company we use next structure:

```
├── cartridges <-- contains all cartridges
├── dw.json
├── dw.json.example
├── gulpfile.js
├── node_modules
├── package.json
├── package-lock.json
├── README.md
├── ssl
└── types
```
Hence need the better way to detect that case.

PS. Also, would be great considered `modules` folder as cartridge too.

**Comment by SqrTT** [comment]
Great to see someone input. I'll take a look.

File: src/providers/CartridgesView.ts (hunk 1 above)
Comment (taurgis) [inline src/providers/CartridgesView.ts]: A possible solution to find the cartridges, no matter where they are is to look for all .project files with 

`	<natures>
		<nature>com.demandware.studio.core.beehiveNature</nature>
	</natures>
`

**Comment by taurgis** [comment]
Used package "find" to look for .project files with a regex check on the contents to see if its a Salesforce (demandware) cartridge.

**Comment by taurgis** [comment]
Added:
 - Look in entire workspace for cartridges (no matter where they are)
 - Support modules
 - Added command to refresh the treeview
 - Added command to create a cartridge

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 * Added quick open for custom tags
 * Added `extension.prophet.cartridges.path` property that allows quick open don't ask a user to choose the file.
-* Added `extension.prophet.ismlServer.activateOn` property that allow activate isml server for non standatd (isml) files, ex. `html`
+* Added `extension.prophet.ismlServer.activateOn` property that allow activate isml server for non standar1d (isml) files, ex. `html`
```
Comment (SqrTT) [inline CHANGELOG.md]: `standar1d` -> `standatd`

File: package.json [hunk 3]
```diff
@@ -28,6 +28,7 @@
     },
     "dependencies": {
         "chokidar": "^1.7.0",
+        "find": "^0.2.7",
```
Comment (SqrTT) [inline package.json]: Is not `glob` solve the same issue?

File: src/providers/CartridgesView.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+	private getCartridgesInWorkspace(workspaceRoot: string): CartridgeItem[] {
+		if (pathExists(workspaceRoot)) {
+			var projectFiles = find.fileSync(/\.project/, workspaceRoot);
+
+			const checkIfCartridge = (projectFile: string): boolean => {
+				let fileContent = fs.readFileSync(projectFile, 'UTF-8');
+
+				// Use a regex check rather than parsing the XML since the file is not that big.
+				var nature = fileContent.match(/<nature>com.demandware.studio.core.beehiveNature<\/nature>/);
```
Comment (SqrTT) [inline src/providers/CartridgesView.ts]: Probably `fileContent.includes('com.demandware.studio.core.beehiveNature')` would be faster

File: src/lib/FileHelper.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import * as path from 'path';
+
+export function getDirectories(srcpath) {
+    return fs.readdirSync(srcpath)
+        .filter(file => fs.lstatSync(path.join(srcpath, file)).isDirectory())
+}
+
+export function getFiles(srcpath) {
+    return fs.readdirSync(srcpath)
+        .filter(file => !fs.lstatSync(path.join(srcpath, file)).isDirectory())
```
Comment (SqrTT) [inline src/lib/FileHelper.ts]: `isFile`?

File: src/lib/FileHelper.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+'use strict';
```
Comment (SqrTT) [inline src/lib/FileHelper.ts]: Is there a way to do this job asynchronously?

File: src/providers/CartridgesView.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+			return directories.map(toFolderElement).concat(files.map(toFileElement));
+		}
+
+		return [new CartridgeItem('No files', 'cartridge-file', '', vscode.TreeItemCollapsibleState.None)];
+	}
+
+
+	private getCartridgesInWorkspace(workspaceRoot: string): CartridgeItem[] {
+		if (pathExists(workspaceRoot)) {
+			var projectFiles = find.fileSync(/\.project/, workspaceRoot);
```
Comment (SqrTT) [inline src/providers/CartridgesView.ts]: Basically, we can find out cartridges root based on `dw.json` and it's config.

File: package.json (hunk 3 above)
Comment (taurgis) [inline package.json]: Yep it does, made the required changes. Didn't notice that glob dit something similar (haven't used the package before).

File: src/providers/CartridgesView.ts (hunk 7 above)
Comment (taurgis) [inline src/providers/CartridgesView.ts]: Yes, but only if it exists. Currently its only required if you want to make use of the upload feature.

File: src/lib/FileHelper.ts (hunk 5 above)
Comment (taurgis) [inline src/lib/FileHelper.ts]: Changed!

File: src/lib/FileHelper.ts (hunk 6 above)
Comment (taurgis) [inline src/lib/FileHelper.ts]: Which job do you mean?

Comment (taurgis) [inline src/lib/FileHelper.ts]: I have made the getfiles and getdirectories an async function using promises.

**Pull request merged**